### ISSUE TITLE
Rich text copy for code editor

### ISSUE THREAD

**Issue by jqrun** [jupyterlab/jupyterlab#9657]
### Problem

I would like to be able to copy code from the code editor into Outlook for example and retain the Codemirror formatting (color, font-style, font-weight, etc) as inline styles. 

### Proposed Solution

Ideally this is exposed as an api in `@jupyterlab/codemirror` that can be used in extensions. Additionally, it would be useful to be able to toggle whether or not to include formatting changes like selection and search.

### Additional context

It's somewhat possible right now by using computed styles, extrapolating from the structure of `@jupyterlab/codemirror` stylesheet variable structures, and the token types from the exposed Codemirror editor/doc. Though this isn't ideal as it's a bit hacky. As an added bonus, Jupyterlab could use this as well to implement native copy paste with rich formatting like other editors (VSCode for example).

Thanks!

**Comment by jasongrout** [comment]
> like other editors (VSCode for example).

VSCode copies code with formatting? How do you that?

As an initial thought, this sounds like it would be hard to do well.

**Comment by jqrun** [comment]
> > like other editors (VSCode for example).
> 
> VSCode copies code with formatting? How do you that?
> 
> As an initial thought, this sounds like it would be hard to do well.

Yes, a reproducer for that would be to copy any arbitrary code to clipboard in VSCode, and then paste it to an Outlook email. The result is code block complete with syntax highlighting.

For starters here, it would at least be useful to expose a basic html construct that applies styles per Codemirror token type since the styles are already organized per token type: https://github.com/jupyterlab/jupyterlab/blob/master/packages/codemirror/style/base.css

```
html = ''
for line in codemirrorline:
  for token in line:
    if token.type:
      html += <span ${styles(token.type)}>
    else:
      html += <span>
    
    html += token.string + </span>
  html += '<br>'
```

**Comment by krassowski** [comment]
See solutions here: https://discuss.codemirror.net/t/rich-text-copy/695/4

**Comment by jqrun** [comment]
> See solutions here: https://discuss.codemirror.net/t/rich-text-copy/695/4

Ah yeah I have already seen the discussions on the codemirror forums, but basically all existing hacks there require computing styles/scraping from stylesheets (and those actually wouldn't successfully ignore overlays from search/selection).

It's definitely already possible to do via an extension that makes some assumptions about Jupyterlab's styling structure and use of `getComputedStyle` + `getPropertyValue`.But I think that it would be much more clean and robust if `@jupyterlab/codemirror` offered this as an API, since it already owns exactly the styling for each token type. Also I think people would be very happy if they could natively copy paste input/output from lab with simple formatting (monospace, colors, text styles).

**Comment by krassowski** [comment]
Implementation wise it could completely ignore CodeMirror existence and use the Markdown render, rendering the code wrapped with tripplebackticks (with proper language mode) to an off-screen HTML element and then just copying the generated HTML to clipboard. It would be much cleaner I think, and would avoid any overlay issues. Also, it should be easily possible to implement with the current JupyterLab public APIs.

**Comment by jqrun** [comment]
> Good point!
> 
> Implementation wise it could completely ignore CodeMirror existence and use the Markdown render, rendering the code wrapped with triple backticks (with proper language mode) to an off-screen HTML element and then just copying the generated HTML to clipboard. It would be much cleaner I think, and would avoid any overlay issues. Also, it should be easily possible to implement with the current JupyterLab public APIs.

Interesting idea with the markdown render off screen! Could you point me to which renderer to use for that? Although when looking at the output of some markdown code, it still uses the cm-${token.type} classes in the HTML just like the code editor, so you would still have to use computed styles to place them inline for the clipboard.

**Comment by krassowski** [comment]
I was thinking about using `@jupyterlab/rendermime` as for example here:

https://github.com/jupyterlab/jupyterlab/blob/b4db7f03d2ad0b91a9c8c252c56ae5e4a9408fbb/packages/tooltip/src/widget.ts#L77-L78

> Although when looking at the output of some markdown code, it still uses the cm-${token.type} classes in the HTML just like the code editor, so you would still have to use computed styles to place them inline for the clipboard.

Maybe. On one hand browsers are intelligent enough to be able to copy the relevant styles when it is a user-invoked copy operation, but on the other I just tested it with a code snippet in markdown in Chrome and see that bold style was correctly copied but the colors were effectively lost as those are set as CSS variables (well the style was inlined properly copied but the referenced variables are no longer there after paste into a different document); this would be browser dependent I believe.

**Comment by jqrun** [comment]
Ah yeah the CSS variables seems to be what's lost in translation. Had someone also mention that when copying output text to Outlook directly, the font family that shows up in the editor there is something like var(--jp-code-font-family). I wonder if there is an easy way to change the behavior where the clipboard takes the computed css variable instead of the reference, that would probably give a lot of free functionality to any client of the paste. Or maybe some css processing to convert to static values?

**Comment by jqrun** [comment]
Just want to update here and add that the rendermime solution suggested by @krassowski works well and better than what I was using previously. When inferring based on token.type, it became an issue when that type field contained a space separated list of types.

For extra reference, I used `renderMarkdown` from `@jupyterlab/rendermime`, hosted in a hidden div. Then computed styles were inlined recursively with postorder DFS, only pulling font color, weight, and style for simplicity.
